File: python-algorithm/ds_5_tree/tree_1_1_bool_recurrent/L572.py
#--coding:utf-8--
# Definition for a binary tree node.

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSubtree(self, root: TreeNode, subRoot: TreeNode) -> bool:
        if not root: return False
        if self.isSame(root, subRoot):
            return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

    def isSame(self, r1, r2):
        if not r1 and not r2:
            return True
        if not r1 or not r2:
            return False
        if r1.val != r2.val:
            return False
        return self.isSame(r1.left, r2.left) and self.isSame(r1.right, r2.right)

File: python-algorithm/ds_5_tree/tree_1_1_bool_recurrent/L572.cpp


#include <iostream>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isSubtree(TreeNode* s, TreeNode* t) {
        if(s == NULL || t == NULL) return false;
        return DoesTree1HaveTree2(s,t)||isSubtree(s->left, t)||isSubtree(s->right, t);
        // bool result = false ;
        // if(s->val == t->val )
        //     result = DoesTree1HaveTree2(s,t);
        // if(!result)
        //     result = isSubtree(s->left, t);
        // if(!result)
        //     result = isSubtree(s->right, t);
        // return result ;
    }

    //判断2棵树是否相等
    bool DoesTree1HaveTree2(TreeNode* pRoot1, TreeNode* pRoot2){
        if(pRoot1==NULL && pRoot2==NULL)  //两个都是空
            return true ;
        if(pRoot1==NULL || pRoot2 == NULL) //有一个为空
            return false ;
        if(pRoot1->val != pRoot2->val )//都不为空，且值不相等
            return false;
        return DoesTree1HaveTree2(pRoot1->left, pRoot2->left) &&
               DoesTree1HaveTree2(pRoot1->right, pRoot2->right);
    }

};